### PR TITLE
[SemiAuto] add static branch for shard_tensor

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR types

New features
### PR changes

APIs
### Description

add static branch for shard_tensor and support dy2static
Pcard-73145

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/api.py [hunk 1]
```diff
@@ -55,15 +58,46 @@ def __init__(self, mesh, sharding_specs):
...
+        self._process_mesh = mesh
+        self._dims_mapping = dims_mapping
+
+    @property
+    def process_mesh(self):
+        """
+        Get process_mesh of the dist_attr
+
+        Returns:
+            paddle.distributed.ProcessMesh: process_mesh
```
Comment (chenwhql) [inline python/paddle/distributed/auto_parallel/api.py]: 还需要补充下示例吗

Comment (zhiqiu) [reply python/paddle/distributed/auto_parallel/api.py]: 感觉不加也可以，比较简单的property函数。如果需要，后续我补一下。

File: python/paddle/distributed/auto_parallel/api.py [hunk 2]
```diff
@@ -55,15 +58,46 @@ def __init__(self, mesh, sharding_specs):
...
         dims_mapping = [
             mesh.dim_names.index(dim_name) if dim_name is not None else -1
             for dim_name in sharding_specs
         ]
 
         # 2. init core.TensorDistAttr
         core.TensorDistAttr.__init__(self)
-        self.process_mesh = mesh
-        self.dims_mapping = dims_mapping
+        self._process_mesh = mesh
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/api.py]: the process_mesh field in c++ TensorDistAttr might remain unchanged ?

Comment (zhiqiu) [reply python/paddle/distributed/auto_parallel/api.py]: fixed

**Review by zhangbo9674 — APPROVED** [review]

**Review by chenwhql — APPROVED** [review]
LGTM

**Pull request merged**